fwdctl/src/dump.rs

```diff
@@ -24,50 +28,230 @@ pub struct Options {
-                } else {
-                    println!("'{}': '{}'", u8_to_string(&key), u8_to_string(&value));
+    let mut output_handler = create_output_handler(opts).expect("Error creating output handler");
+
+    while let Some(item) = stream.next().await {
+        match item {
+            Ok((key, value)) => {
+                output_handler.handle_record(&key, &value);
+
+                key_count += 1;
+
+                if key == stop_key || key_count_exceeded(opts.max_key_count, key_count) {
+                    handle_next_key(stream.next().await).await;
+                    break;
                 }
             }
-            Some(Err(e)) => return Err(e),
+            Err(e) => return Err(e),
         }
     }
+    output_handler.flush();
+
     Ok(())
 }
 
-fn u8_to_string(data: &[u8]) -> Cow<'_, str> {
-    String::from_utf8_lossy(data)
+const fn key_count_exceeded(max: Option<u32>, key_count: u32) -> bool {
+    matches!(max, Some(max) if key_count >= max)
+}
+
+fn u8_to_string(data: &[u8]) -> String {
+    String::from_utf8_lossy(data).to_string()
 }
 
 fn key_parser(s: &str) -> Result<Box<[u8]>, std::io::Error> {
     Ok(Box::from(s.as_bytes()))
 }
+
+// Helper function to convert key and value to a string
+fn key_value_to_string(key: &[u8], value: &[u8], hex: bool) -> (String, String) {
+    let key_str = if hex {
+        hex::encode(key)
+    } else {
+        u8_to_string(key)
+    };
+    let value_str = if hex {
+        hex::encode(value)
+    } else {
+        u8_to_string(value)
+    };
+    (key_str, value_str)
+}
+
+async fn handle_next_key(next_key: KeyFromStream) {
+    match next_key {
+        Some(Ok((key, _))) => {
+            println!(
+                "Next key is {0}, resume with \"--start-key={0}\".",
+                u8_to_string(&key)
+            );
+        }
+        Some(Err(e)) => {
+            eprintln!("Error occurred while fetching the next key: {}", e);
+        }
+        None => {
+            println!("There is no next key. Data dump completed.");
+        }
+    }
+}
+
+trait OutputHandler {
+    fn handle_record(&mut self, key: &[u8], value: &[u8]);
+    fn flush(&mut self);
+}
+
+struct CsvOutputHandler {
+    writer: csv::Writer<File>,
+    hex: bool,
+}
+
+impl OutputHandler for CsvOutputHandler {
+    fn handle_record(&mut self, key: &[u8], value: &[u8]) {
+        let (key_str, value_str) = key_value_to_string(key, value, self.hex);
+        self.writer
+            .write_record(&[key_str, value_str])
+            .expect("Failed to write CSV record");
+    }
+
+    fn flush(&mut self) {
+        self.writer.flush().expect("Failed to flush CSV writer");
+    }
+}
+
+struct JsonOutputHandler {
+    writer: BufWriter<File>,
+    hex: bool,
+    is_first: bool,
+}
+
+impl JsonOutputHandler {
+    fn new(mut writer: BufWriter<File>, hex: bool) -> JsonOutputHandler {
+        let _ = writer.write("{\n".as_bytes());
```

Comment:
```suggestion
    writer.write(b"{\n")
```

There are a number of other places you can avoid calling `as_bytes()`